File: Simple/Multiply.rb
class Multiply < Struct.new(:left, :right)
  def reduce
    if left.reducible?
      Multiply.new(left.reduce, right)
    elsif right.reducible?
      Multiply.new(left, right.reduce)
    else
      Number.new(left.value * right.value)
    end
  end

  def to_s
    "#{left} * #{right}"
  end

  def inspect
    "«#{self}»"
  end

  def reducible?
    true
  end
end

File: chapter02/BigStepSemantic/Multiply.py
from chapter02.BigStepSemantic.Number import Number


class Multiply(object):
    def __init__(self, left, right):
        self.left = left
        self.right = right

    def evaluate(self, environment):
        return Number(self.left.evaluate(environment).value * self.right.evaluate(environment).value)
